extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/photos/MicrosoftPhotosImporter.java

```diff
@@ -176,86 +179,51 @@ private String createOneDriveFolder(PhotoAlbum album) throws IOException {
+    } else {
+      throw new IllegalStateException("Don't know how to get the inputStream for " + photo);
+    }
 
-      if (photo.isInTempStore()) {
-        inputStream = jobStore.getStream(jobId, photo.getFetchableUrl()).getStream();
-      } else if (photo.getFetchableUrl() != null) {
-        inputStream = new URL(photo.getFetchableUrl()).openStream();
-      } else {
-        throw new IllegalStateException("Don't know how to get the inputStream for " + photo);
-      }
+    String photoUploadUrl = createUploadSession(photo, idempotentImportExecutor);
+
+    // Arrange the data to be uploaded in chunks
+    List<DataChunk> chunksToSend = DataChunk.splitData(inputStream);
+    final int totalFileSize = chunksToSend.stream().map(DataChunk::getSize).reduce(0, Integer::sum);
 
-      Request.Builder requestBuilder = new Request.Builder().url(uploadUrl);
-      requestBuilder.header("Authorization", "Bearer " + credential.getAccessToken());
-
-      MediaType contentType = MediaType.parse(photo.getMediaType());
-
-      StreamingBody body = new StreamingBody(contentType, inputStream);
-
-      // PUT the stream
-      requestBuilder.put(body);
-
-      try (Response response = client.newCall(requestBuilder.build()).execute()) {
-        int code = response.code();
-        ResponseBody responseBody = response.body();
-        if (code == 401) {
-          // If there was an unauthorized error, then try refreshing the creds
-          credentialFactory.refreshCredential(credential);
-          monitor.info(() -> "Refreshed authorization token successfuly");
-
-          requestBuilder.header("Authorization", "Bearer " + credential.getAccessToken());
-          Response newResponse = client.newCall(requestBuilder.build()).execute();
-          code = newResponse.code();
-          responseBody = newResponse.body();
-        }
-        if (code < 200 || code > 299) {
-          throw new IOException(
-              "Got error code: "
-                  + code
-                  + " message: "
-                  + response.message()
-                  + " body: "
-                  + response.body().string());
-        }
-
-        // Extract photo ID from response body
-        Preconditions.checkState(body != null, "Got Null Body when creating photo %s", photo);
-        Map<String, Object> responseData = objectMapper.readValue(responseBody.bytes(), Map.class);
-        String photoId = (String) responseData.get("id");
-        checkState(
-            !Strings.isNullOrEmpty(photoId), "Expected id value to be present in %s", responseData);
-        return photoId;
+    Integer chunkCode = null;
+    Response chunkResponse = null;
+    for (DataChunk chunk : chunksToSend) {
+      chunkResponse = uploadChunk(chunk, photoUploadUrl, totalFileSize, photo.getMediaType());
+      chunkCode = chunkResponse.code();
+      if (chunkCode < 200 || chunkCode > 299) {
```

Comment:
I might move this exception detection stuff into the uploadChunk method, around where you're doing the 401 catch.